src/prng/chacha.rs

```diff
@@ -292,10 +248,96 @@ impl SeedableRng for ChaChaCore {
 
 impl CryptoRng for ChaChaCore {}
 
+/// A builder type allowing extra parameterisation of ChaCha.
+///
+/// Construct using `SeedableRng` as normal.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::{Rng, SeedableRng};
+/// use rand::prng::chacha::{ChaChaRng, ChaChaBuilder};
+///
+/// let mut rng: ChaChaRng =
+///     ChaChaBuilder::from_seed([0u8; 32])
+///         .set_rounds(12)
+///         .set_stream(123456)
+///         .into();
+/// rng.gen::<u32>();
+/// ```
+#[derive(Clone, Debug)]
+pub struct ChaChaBuilder {
+    core: ChaChaCore,
+}
+
+impl ChaChaBuilder {
```

Comment:
I don't really like the builder approach here. Does it have any real advantages?

The three provided methods here also make sense an the RNG itself, not only on the builder.

Reply:
Actually, there's a little problem with setting the nonce: one may need to regenerate the current buffer, and it doesn't exactly make sense to do that mid-buffer — surely you'd only set the nonce if resetting the counter too? When it comes to selecting a stream I'm not sure.

The `set_rounds` thing I don't know; I'm still not quite convinced what we should do with this especially with regards to `CryptoRng`.

Reply:
I am (already was) ok with hiding/removing `set_rounds` for now. Actually I glanced over the function too fast, I though the three builder methods for for nonce, stream and counter. What do you think of just three 'setting' methods on `ChaChaRng`, like so?
- `set_position(&mut self, position: u128)`. High-level, has to regenerate the buffered results if the position does not lay on a block boundary. Otherwise waits with generating until on next use. Position is the position of the next value to return.
- `set_stream(&mut self, stream: u64)`. High-level, schedules generating a new block of results.
- `set_counter_and_nonce(&mut self, counter: u64, nonce: [u8; 8])`. Low-level, also schedules generating a new block of results.

And I suppose three corresponding getters.

Reply:
The reset does nothing more than reset the index to `results.len()`. If this is optimized out or not is immeasurable.

I thought combining them made sense because as you said, if you change the nonce you probably also want to change the counter. And if you want to change only the counter, either do it in combination with the nonce or use `set_position` (when `i128` gets more common).